### PR TITLE
Add support for writing updateStatus to the status blob if requested version in GS

### PR DESCRIPTION
**Pull request by larohra**
## Description
This PR adds support for writing updateStatus to the status blob if requested version in GS. If not, we don't report anything for the updateStatus.

Issue # 

---

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] If applicable, the PR references the bug/issue that it fixes in the description.
- [ ] New Unit tests were added for the changes made

### Quality of Code and Contribution Guidelines
- [ ] I have read the [contribution guidelines](https://github.com/Azure/WALinuxAgent/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: azurelinuxagent/common/protocol/restapi.py [hunk 1]
```diff
@@ -68,9 +72,21 @@ class VMAgentManifest(object):
     def __init__(self, family, version=None):
         self.family = family
         # This is the Requested version as specified by the Goal State, it defaults to 0.0.0.0 if not specified in GS
```
Comment (kevinclark19a) [inline azurelinuxagent/common/protocol/restapi.py:74]: Should this comment be moved up to where `VERSION_0` is defined?

Comment (larohra) [reply azurelinuxagent/common/protocol/restapi.py:74]: `VERSION_0` is just a constant depicting the `0.0.0.0` value. This comment is specific to the logic specified here so I feel its in the apt place.

File: azurelinuxagent/ga/exthandlers.py [hunk 2]
```diff
@@ -889,15 +889,15 @@ def __get_handlers_on_file_system(self, incarnation_changed):
...
         """
         Go through handler_state dir, collect and report status.
         Returns the status it reported, or None if an error occurred.
         """
         try:
             vm_status = VMStatus(status="Ready", message="Guest Agent is running",
-                                 gs_aggregate_status=self.__gs_aggregate_status)
-
+                                 gs_aggregate_status=self.__gs_aggregate_status,
+                                 vm_agent_update_status=vm_agent_upgrade_status)
```
Comment (kevinclark19a) [inline azurelinuxagent/ga/exthandlers.py:900]: NIT: these two have slightly different names (`...agent_update...` vs `...agent_upgrade...`), might consider picking just a single one.

Comment (larohra) [reply azurelinuxagent/ga/exthandlers.py:900]: good catch! Let me fix it.

File: azurelinuxagent/common/protocol/restapi.py [hunk 3]
```diff
@@ -68,9 +72,21 @@ class VMAgentManifest(object):
...
-        self.version = "0.0.0.0" if version is None else version
+        self.version = VERSION_0 if version is None else version
         self.uris = []
 
+    @property
+    def requested_version(self):
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/restapi.py:79]: I think having a requested_version and a version properties is confusing... maybe the latter should be called requested_version_string?

File: azurelinuxagent/common/protocol/restapi.py [hunk 4]
```diff
@@ -68,9 +72,21 @@ class VMAgentManifest(object):
...
+        return FlexibleVersion(self.version)
+
+    @property
+    def is_requested_version_available(self):
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/restapi.py:83]: "available" may not be the correct term... "specified"?

File: azurelinuxagent/common/protocol/restapi.py [hunk 5]
```diff
@@ -68,9 +72,21 @@ class VMAgentManifest(object):
...
+        """
+        If we don't get any requested_version from the GS, we default it to 0.0.0.0.
```
Comment (narrieta) [inline azurelinuxagent/common/protocol/restapi.py:85]: doesn't the spec mention that the requested version is always in the GS?

File: azurelinuxagent/ga/update.py [hunk 6]
```diff
@@ -487,9 +488,44 @@ def _process_goal_state(self, exthandlers_handler, remote_access_handler):
...
+        """
+        This function gets the VMAgent upgrade status as per the last GoalState.
+        Returns: None if the last GS does not ask for requested version else VMAgentUpdateStatus
+        """
+        upgrade_status = None
+
+        try:
+            agent_manifests, _ = protocol.get_vmagent_manifests()
+            # Expectation here is that there will only be one manifest per family passed down from CRP.
+            # If there are multiple, we pick the first one
```
Comment (narrieta) [inline azurelinuxagent/ga/update.py:502]: what would multiple statuses indicate? an error? do we need to log that?

File: tests/ga/test_update.py [hunk 7]
```diff
@@ -1740,6 +1741,66 @@ def test_it_should_recreate_extension_event_directories_for_existing_extensions_
...
 
+    def test_it_should_report_update_status_in_status_blob(self):
+        with _get_update_handler(iterations=1) as (update_handler, protocol):
+            with patch.object(conf, "get_autoupdate_gafamily", return_value="Prod"):
+                with patch("azurelinuxagent.common.logger.warn") as patch_warn:
+
+                    protocol.aggregate_status = None
+                    protocol.incarnation = 1
+
+                    def mock_http_put(url, *args, **_):
```
Comment (kevinclark19a) [inline tests/ga/test_update.py:1752]: this feels like it could break if different types of PUT calls are added to the agent; there are only two choices below and one of them is assumed (i.e. not explicitly checked for, it's just in  the equivalent of an `else` clause.

I think this could be more or less fixed by making the second scenario explicit (via an if statement) and then do a `return None` to ignore anything else.

File: tests/ga/test_update.py [hunk 8]
```diff
@@ -1740,6 +1741,66 @@ def test_it_should_recreate_extension_event_directories_for_existing_extensions_
...
+                        self.assertTrue(update_handler.exit_mock.called, "The process should have exited")
+                        exit_args, _ = update_handler.exit_mock.call_args
+                        self.assertEqual(exit_args[0], 0,
+                                         "Exit code should be 0; List of all warnings logged by the agent: {0}".format(
+                                             patch_warn.call_args_list))
+
+                    protocol.set_http_handlers(http_put_handler=mock_http_put)
+
+                    # Case 1: No requested version in GS; updateStatus should not be reported
+                    update_handler.run(debug=True)
```
Comment (kevinclark19a) [inline tests/ga/test_update.py:1774]: A bit confused by this.. Why not use the `update_goal_state_and_run_handler` method from above?

File: tests/ga/test_update.py [hunk 9]
```diff
@@ -1740,6 +1741,66 @@ def test_it_should_recreate_extension_event_directories_for_existing_extensions_
...
+                    self.assertFalse("updateStatus" in protocol.aggregate_status['aggregateStatus']['guestAgentStatus'],
+                                     "updateStatus should not be reported if not asked in GS")
+
+                    # Case 2: Requested version in GS != Current Version; updateStatus should not be error
```
Comment (kevinclark19a) [inline tests/ga/test_update.py:1778]: this comment seems backwards... Case 2 is expecting an error, right?
```python
self.assertEqual(VMAgentUpdateStatuses.Error, update_status['status'], "Status should be an error")
```

File: azurelinuxagent/ga/update.py [hunk 10]
```diff
@@ -487,9 +488,44 @@ def _process_goal_state(self, exthandlers_handler, remote_access_handler):
...
+                err_msg = "[This error will only be logged once per incarnation] " \
+                          "Ran into error when trying to fetch updateStatus for the agent, skipping reporting update satus. Error: {0}".format(
+                           textutil.format_exception(error))
+                logger.warn(err_msg)
+                add_event(op=WALAEventOperation.AgentUpgrade, is_success=False, message=err_msg, log_event=False)
+
+        return upgrade_status
+
     def _report_status(self, exthandlers_handler, incarnation_changed):
+        vm_agent_upgrade_status = self.__get_vmagent_upgrade_status(exthandlers_handler.protocol, incarnation_changed)
```
Comment (narrieta) [inline azurelinuxagent/ga/update.py:525]: should this be disabled in production until the feature is stable?

File: azurelinuxagent/ga/update.py [hunk 11]
```diff
@@ -487,9 +488,44 @@ def _process_goal_state(self, exthandlers_handler, remote_access_handler):
...
+            manifest = next(m for m in agent_manifests if m.family == conf.get_autoupdate_gafamily())
+            if manifest.is_requested_version_available:
+                if CURRENT_VERSION == manifest.requested_version:
+                    status = VMAgentUpdateStatuses.Success
+                    code = 0
+                else:
+                    status = VMAgentUpdateStatuses.Error
+                    code = -1
+                upgrade_status = VMAgentUpdateStatus(expected_version=manifest.version, status=status, code=code)
+        except StopIteration:
```
Comment (narrieta) [inline azurelinuxagent/ga/update.py:512]: could we have this except clause enclose only the actual iteration? if at some point another iteration is added to this code this could lead to problems

File: azurelinuxagent/ga/update.py [hunk 12]
```diff
@@ -487,9 +488,44 @@ def _process_goal_state(self, exthandlers_handler, remote_access_handler):
...
+            logger.verbose("No matching manifest found for family: {0}".format(conf.get_autoupdate_gafamily()))
```
Comment (narrieta) [inline azurelinuxagent/ga/update.py:513]: should we log this one also once per goal state?

File: tests/ga/test_update.py [hunk 13]
```diff
@@ -1740,6 +1741,66 @@ def test_it_should_recreate_extension_event_directories_for_existing_extensions_
...
+                    protocol.mock_wire_data.set_extension_config("wire/ext_conf_requested_version.xml")
+                    update_goal_state_and_run_handler()
+                    self.assertTrue("updateStatus" in protocol.aggregate_status['aggregateStatus']['guestAgentStatus'],
+                                    "updateStatus should be in status blob. Warns: {0}".format(patch_warn.call_args_list))
+                    update_status = protocol.aggregate_status['aggregateStatus']['guestAgentStatus']["updateStatus"]
+                    self.assertEqual(VMAgentUpdateStatuses.Error, update_status['status'], "Status should be an error")
+                    self.assertEqual(update_status['expectedVersion'], "9.9.9.10", "incorrect version reported")
+                    self.assertEqual(update_status['code'], -1, "incorrect code reported")
+
+                    # Case 3: Requested version in GS == Current Version; updateStatus should not be Success
```
Comment (kevinclark19a) [inline tests/ga/test_update.py:1788]: Same as above, isn't the code below expecting success?

File: azurelinuxagent/common/protocol/wire.py [hunk 14]
```diff
@@ -292,9 +292,23 @@ def ga_status_to_v1(ga_status):
...
         "formattedMessage": __get_formatted_msg_for_status_reporting(ga_status.message)
     }
+
+    if ga_status.update_status is not None:
+        v1_ga_status["updateStatus"] = get_ga_update_status_to_v1(ga_status.update_status)
+
     return v1_ga_status
 
 
+def get_ga_update_status_to_v1(update_status):
```
[the remark this answers is not in the document]
Comment (kevinclark19a) [reply azurelinuxagent/common/protocol/wire.py:302]: +1

File: azurelinuxagent/ga/update.py [hunk 15]
```diff
@@ -487,9 +488,44 @@ def _process_goal_state(self, exthandlers_handler, remote_access_handler):
...
+
+        try:
+            agent_manifests, _ = protocol.get_vmagent_manifests()
+            # Expectation here is that there will only be one manifest per family passed down from CRP.
+            # If there are multiple, we pick the first one
+            manifest = next(m for m in agent_manifests if m.family == conf.get_autoupdate_gafamily())
+            if manifest.is_requested_version_available:
+                if CURRENT_VERSION == manifest.requested_version:
+                    status = VMAgentUpdateStatuses.Success
+                    code = 0
```
[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/ga/update.py:507]: Its used for logging in CRP. Windows agent primarily sends different HR Codes for different cases which gets logged in CRP. Here's what Aymen said about it - 

> Usually it would be an error code(mainly for windows)
Its only for logging
CRP will only cares about one thing : "status"

File: azurelinuxagent/ga/update.py (hunk 11 above)
Comment (larohra) [reply azurelinuxagent/ga/update.py:512]: Done

File: azurelinuxagent/ga/update.py (hunk 12 above)
Comment (larohra) [reply azurelinuxagent/ga/update.py:513]: Done

File: azurelinuxagent/ga/update.py (hunk 10 above)
Comment (larohra) [reply azurelinuxagent/ga/update.py:525]: This would only be reported if CRP includes a requested version in GS. This does no harm to the agent or CRP since its just adding to the status blob. Even if by a bug CRP sends us a requested version in GS before the feature is stable and we end up adding the `updateStatus` in the status blob, it wont do us any harm. It'll just be ignored.

Did you have any other use case in mind?

File: tests/ga/test_update.py (hunk 7 above)
Comment (larohra) [reply tests/ga/test_update.py:1752]: That's a good point, we should indeed make it explicit. That being said, unfortunately, as per the current code, its very hard to differentiate if a PUT call is for status or not because we directly upload to the storage blob uri we get from GS - 

> if self.status_blob.upload(extensions_goal_state.status_upload_blob):

The only other differentiating factor we have is a PUT call. One way to deal with it would be to change all of our test files and add a unique string in the `<StatusUploadBlob>` object to specify that its a status blob and then filter it out based on those, but for the time being, since we dont have any other PUT calls, I'd rather keep it as is to avoid the extra trouble :)

File: tests/ga/test_update.py (hunk 8 above)
Comment (larohra) [reply tests/ga/test_update.py:1774]: No particular reason honestly, for the subsequent calls we needed these functions - 

> protocol.mock_wire_data.set_incarnation(protocol.incarnation)
                        self._add_write_permission_to_goal_state_files()
                        update_handler.set_iterations(1)

But for the first one it wasn't needed so skipped it. But since you pointed it out, it is beneficial to call the `update_goal_state_and_run_handler` here too since it has an extra assert so added it here

File: tests/ga/test_update.py (hunk 9 above)
Comment (larohra) [reply tests/ga/test_update.py:1778]: Good catch! I think I got confused between the wording of case 2 and 3. Fixed it thanks!

File: tests/ga/test_update.py (hunk 13 above)
Comment (larohra) [reply tests/ga/test_update.py:1788]: Fixed

File: tests/ga/test_update.py (hunk 7 above)
Comment (kevinclark19a) [reply tests/ga/test_update.py:1752]: Could you just create a new GS file to use for this test and edit it such that the storage blob uri matches some regex (or just straight up is some string) and filter off that?

That would still be a bit fragile (i.e. someone changing the specific file you used might break things) but I feel like it's a good compromise between leaving it as is and doing the extra work of manipulating the status blob string itself.

**Review by kevinclark19a — APPROVED** [review]

File: tests/ga/test_update.py (hunk 7 above)
Comment (larohra) [reply tests/ga/test_update.py:1752]: After thinking about this again, since its being used in multiple places, I'd rather fix this in a separate PR in all those usages.

File: azurelinuxagent/common/protocol/wire.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/common/protocol/wire.py:302]: thanks; i guess we can leave it as is

File: azurelinuxagent/ga/update.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/ga/update.py:515]: if crp does not care about it, why populate it at all? or we need some documentation about what 0 and -1 mean

File: azurelinuxagent/ga/update.py (hunk 6 above)
Comment (narrieta) [reply azurelinuxagent/ga/update.py]: maybe change the comment to say that during validation we already checked there is only 1 item so we just pick up the first one?

File: azurelinuxagent/ga/update.py (hunk 10 above)
Comment (narrieta) [reply azurelinuxagent/ga/update.py]: where do we ensure that we report this only if the GS includes a requested version?

File: azurelinuxagent/ga/update.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/ga/update.py:515]: CRP logs it. I think the intention is to have a clear message on what went wrong in the CRP logs itself. 

>  why populate it at all?

It's primarily needed for Windows where they're going to return different HR codes based on the scenario.   

> or we need some documentation about what 0 and -1 mean

Since we dont have a specific need for it, I just stuck to the basic exit code usages that we generally use in the code (0 for success and -1 for failure). Let me know if you feel the need to have a comment about it and I can add it.

File: azurelinuxagent/ga/update.py (hunk 10 above)
Comment (larohra) [reply azurelinuxagent/ga/update.py]: The check for this is in Line 512 of update.py - 

```python
if manifest.is_requested_version_specified:
                if CURRENT_VERSION == manifest.requested_version:
                    status = VMAgentUpdateStatuses.Success
                    code = 0
                else:
                    status = VMAgentUpdateStatuses.Error
                    code = -1
                update_status = VMAgentUpdateStatus(expected_version=manifest.requested_version_string, status=status,
                                                    code=code)
```

In case the `manifest.is_requested_version_specified` condition is not satisfied, we return None. 

When we're preparing the status blob, we check if the `vm_status.vmAgent.update_status` is not None and if so, include the status in the blob else skip it (line 296 in wire.py) - 

```py
if ga_status.update_status is not None:
        v1_ga_status["updateStatus"] = get_ga_update_status_to_v1(ga_status.update_status)
```

Hope this helps!

File: azurelinuxagent/ga/update.py (hunk 6 above)
Comment (larohra) [reply azurelinuxagent/ga/update.py]: sure let me rephrase

Comment (larohra) [reply azurelinuxagent/ga/update.py]: Done

File: azurelinuxagent/ga/update.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/ga/update.py:515]: in what instances do we use -1 for failure?

File: azurelinuxagent/ga/update.py (hunk 10 above)
Comment (narrieta) [reply azurelinuxagent/ga/update.py]: I still feel this should be completely disabled in production until the feature is stable. If at some point there is some issue in the feature and we have some agents pushing this status, who knows, that could be an issue.

File: azurelinuxagent/ga/update.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (larohra) [reply azurelinuxagent/ga/update.py:515]: You're right, the normal exit codes are positive non-0s. I'll change it to `1`

File: azurelinuxagent/ga/update.py (hunk 10 above)
Comment (larohra) [reply azurelinuxagent/ga/update.py]: I don't see any big downside to it honestly. It's being controlled by CRP (the availability of `requested_version`) and we would upload status only if asked. On top of it, this status is only meant for RSM so it should be pretty invisible to the customers. 

On the flipside though, if I think about it, one case that comes to mind on when it can break is if CRP messes up and kind of couples the updateStatus with the extension status and fails extensions due to a failed upgrade. I guess we can go with a flagged approach and enable it by default when we have more confidence in the feature e2e. I'll also open a task to add a DCR scenario for it with the flag enabled and another for enabling it globally in the future once its ready.

File: azurelinuxagent/ga/update.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (narrieta) [reply azurelinuxagent/ga/update.py:518]: thanks

**Review by narrieta — APPROVED** [review]

**Review by kevinclark19a — APPROVED** [review]
